### PR TITLE
Hold config lock when modifying load balancers

### PR DESCRIPTION
**Pull request by mpolden**
@bratseth

### CODE REVIEW

**Review by bratseth — APPROVED** [review]
LGTM but I don't see why we need node locks as you mention in the NodeRepository TODO: All changes to nodes are locked by the application lock for allocated nodes and the unprovisioned lock for unallocated nodes.

**Comment by mpolden** [comment]
> I don't see why we need node locks as you mention in the NodeRepository TODO.

Yes, but as far as I can tell the appropriate lock isn't held at the point in time when the transaction is committed. It's only held while appending operations to the transaction. Example: Nodes can be changed at any point after the lock is released here https://github.com/vespa-engine/vespa/blob/master/node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/Activator.java#L44-L47 and before the commit in https://github.com/vespa-engine/vespa/blob/master/configserver/src/main/java/com/yahoo/vespa/config/server/deploy/Deployment.java#L134-L139.

 It's the same problem as with load balancers.

**Comment by bratseth** [comment]
Yes, very good point, but I don't see that we need more fine-grained locks, we just need to share the same application lock on both sides, which is what you achieve here for the load balancer case. Let's just call that the application lock and use it everywhere?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java [hunk 1]
```diff
@@ -761,6 +762,7 @@ else if (node.flavor().getType() == Flavor.Type.DOCKER_CONTAINER) { // removing
     public Zone zone() { return zone; }
 
     /** Create a lock which provides exclusive rights to making changes to the given application */
+    // TODO(mpolden): Make this delegate to CuratorDatabaseClient#lockConfig instead
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java:765]: > Let's just call that the application lock and use it everywhere?

Yes, that's this TODO, but baby steps... :)

**Pull request merged**